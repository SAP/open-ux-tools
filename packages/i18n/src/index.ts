export { getCapI18nBundle, getPropertiesI18nBundle } from './read';
export { createCapI18nEntries, createPropertiesI18nEntries } from './write';

export {
    getI18nFolderNames,
    getI18nMaxLength,
    getI18nTextType,
    extractI18nKey,
    extractDoubleCurlyBracketsKey,
    getI18nUniqueKey,
    convertToCamelCase,
    convertToPascalCase,
    printPropertiesI18nEntry,
    printPropertiesI18nAnnotation
} from './utils';

export {
    CdsEnvironment,
    CdsI18nConfiguration,
    CdsI18nEnv,
    I18nAnnotation,
    I18nAnnotationNode,
    I18nBundle,
    I18nEntry,
    NewI18nEntry,
    SapLongTextType,
    SapShortTextType,
    SapTextType,
    ValueNode
} from './types';

export { csvToI18nBundle, jsonToI18nBundle, propertiesToI18nEntry } from './transformer';
