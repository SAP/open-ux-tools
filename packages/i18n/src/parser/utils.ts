import { Range, Position } from 'vscode-languageserver-types';

export { Range, Position };

/**
 * Computes line offsets for the given string.
 *
 * @param text text
 * @returns array of number
 */
export function getLineOffsets(text: string): number[] {
    const lineOffsets: number[] = [0];
    for (let index = 0; index < text.length; ) {
        const character = text[index];
        if (character === '\n') {
            lineOffsets.push(index + 1);
            index++; // Increment index here
        } else if (character === '\r') {
            if (index + 1 < text.length && text[index + 1] === '\n') {
                index += 2; // Increment index by 2 when encountering '\r\n'
            } else {
                index++; // Increment index by 1 when encountering '\r'
            }
            lineOffsets.push(index); // Push the updated index value
        } else {
            index++; // Increment index if character is not a line break
        }
    }
    return lineOffsets;
}

/**
 * Position at.
 *
 * @param lineOffsets Array of indices with line start offsets.
 * e.g [0] represents a document with one line that starts at offset 0.
 * @param offset offset
 * @param textLength max length
 * @returns position
 */
export function positionAt(lineOffsets: number[], offset: number, textLength: number): Position {
    const target = Math.max(Math.min(offset, textLength), 0);
    let low = 0;
    let high = lineOffsets.length;

    if (high === 0) {
        return Position.create(0, target);
    }

    while (low < high) {
        const mid = Math.floor((low + high) / 2);
        if (lineOffsets[mid] > target) {
            high = mid;
        } else {
            low = mid + 1;
        }
    }
    const line = low - 1;
    return Position.create(line, target - lineOffsets[line]);
}

/**
 * Create range.
 *
 * @param lineOffsets line offset
 * @param start start point
 * @param end end point
 * @param textLength text length
 * @returns range
 */
export function rangeAt(lineOffsets: number[], start: number, end: number, textLength: number): Range {
    return Range.create(positionAt(lineOffsets, start, textLength), positionAt(lineOffsets, end, textLength));
}
