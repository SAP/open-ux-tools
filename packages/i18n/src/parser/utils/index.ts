export { getLineOffsets } from './line-offsets';
export { positionAt } from './position';
export { rangeAt } from './range';
export { Position, Range } from './language-server';
