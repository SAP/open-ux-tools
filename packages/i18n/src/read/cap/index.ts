export { getCapI18nBundle } from './bundle';
