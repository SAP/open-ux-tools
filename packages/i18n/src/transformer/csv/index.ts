export { csvToI18nBundle } from './csv';
