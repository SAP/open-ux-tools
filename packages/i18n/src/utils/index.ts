export { getI18nConfiguration, getI18nFolderNames } from './config';
export { getCapI18nFiles, resolveCapI18nFolderForFile, getCapI18nFolder } from './resolve';
export { csvPath, jsonPath, capPropertiesPath, doesExist } from './path';
export { printPropertiesI18nAnnotation, printPropertiesI18nEntry } from './print';
export {
    getI18nMaxLength,
    getI18nTextType,
    applyIndent,
    discoverIndent,
    discoverLineEnding,
    convertToCamelCase,
    convertToPascalCase
} from './text';
export { extractI18nKey, getI18nUniqueKey, extractDoubleCurlyBracketsKey } from './key';
export { readFile, writeFile } from './mem-fs-editor';
