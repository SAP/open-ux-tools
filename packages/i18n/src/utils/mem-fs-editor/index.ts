export { readFile } from './read';
export { writeFile } from './write';
