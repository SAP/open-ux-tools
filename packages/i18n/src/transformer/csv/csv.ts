import { parseCsv } from '../../parser/csv/parser';
import type { CsvField } from '../../parser/csv/types';
import type { I18nBundle, I18nEntry, TextNode } from '../../types';

/**
 * Convert csv field to text node.
 *
 * @param field csv field
 * @returns text node or undefined
 */
function toTextNode(field: CsvField | undefined): TextNode | undefined {
    if (!field) {
        return undefined;
    }
    return {
        value: field.value,
        range: field.range
    };
}

/**
 * Convert CSV content to i18n bundles.
 *
 * @param text csv text
 * @param filePath file path of csv text
 * @returns i18n bundles
 */
export function csvToI18nBundle(text: string, filePath = ''): I18nBundle {
    const bundle: I18nBundle = {};
    const { ast } = parseCsv(text);
    for (let columnIndex = 1; columnIndex < ast.header.fields.length; columnIndex++) {
        const locale = ast.header.fields[columnIndex];
        const entries: I18nEntry[] = [];
        for (const row of ast.rows) {
            if (row) {
                const key = toTextNode(row.fields[0]);
                const value = toTextNode(row.fields[columnIndex]);
                if (key && value) {
                    entries.push({
                        filePath,
                        key,
                        value
                    });
                }
            }
        }
        bundle[locale.value] = entries;
    }
    return bundle;
}
