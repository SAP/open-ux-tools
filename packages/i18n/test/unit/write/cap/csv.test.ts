import { addCsvTexts, tryAddCsvTexts } from '../../../../src/write/cap/csv';
import * as utils from '../../../../src/utils';
import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';

describe('csv', () => {
    describe('add new i18n entries to csv file', () => {
        test('empty file', () => {
            const result = addCsvTexts('', 'fallback', [
                {
                    key: 'key',
                    value: 'value'
                }
            ]);
            expect(result).toMatchSnapshot();
        });
        test('empty bundle', () => {
            const result = addCsvTexts('key', 'fallback', [
                {
                    key: 'key',
                    value: 'value'
                }
            ]);
            expect(result).toMatchSnapshot();
        });
        test('empty fallback bundle', () => {
            const content = `key;fallback\n`;
            const result = addCsvTexts(content, 'fallback', [
                {
                    key: 'key',
                    value: 'value'
                }
            ]);
            expect(result).toMatchSnapshot();
        });
        test('fallback bundle with multiple values', () => {
            const content = `key;fallback\n`;
            const result = addCsvTexts(content, 'fallback', [
                {
                    key: 'key',
                    value: 'value'
                },
                {
                    key: 'key2',
                    value: 'value2'
                }
            ]);
            expect(result).toMatchSnapshot();
        });
        test('multiple bundles with values', () => {
            const content = `key;en;de;fallback\nkey;e;d;\n`;
            const result = addCsvTexts(content, 'fallback', [
                {
                    key: 'key',
                    value: 'value'
                },
                {
                    key: 'key2',
                    value: 'value2'
                }
            ]);
            expect(result).toMatchSnapshot();
        });
        test('multiple bundles with values, but no fallback bundle', () => {
            const content = `key;en;de\nkey;e;d\n`;
            const result = addCsvTexts(content, 'fallback', [
                {
                    key: 'key',
                    value: 'value'
                },
                {
                    key: 'key2',
                    value: 'value2'
                }
            ]);
            expect(result).toMatchSnapshot();
        });
    });
    describe('tryAddCsvTexts', () => {
        const path = join('root', '_i18n', 'i18n');
        const env = Object.freeze({
            i18n: {
                folders: ['_i18n', 'i18n', 'assets/i18n'],
                default_language: 'en'
            }
        });
        afterEach(() => {
            jest.resetAllMocks();
        });
        const entries = [
            {
                key: 'NewKey',
                value: 'New Value'
            }
        ];
        test('csv file does not exist', async () => {
            // arrange
            const doesExistSpy = jest.spyOn(utils, 'doesExist').mockResolvedValue(false);
            const readFileSpy = jest.spyOn(utils, 'readFile').mockResolvedValue('');
            const writeFileSpy = jest.spyOn(utils, 'writeFile').mockResolvedValue();
            // act
            const result = await tryAddCsvTexts(env, path, entries);
            // assert
            expect(result).toEqual(false);
            expect(doesExistSpy).toHaveBeenCalledTimes(1);
            expect(readFileSpy).toHaveBeenCalledTimes(0);
            expect(writeFileSpy).toHaveBeenCalledTimes(0);
        });
        test('add to existing .csv file', async () => {
            // arrange
            const csvI18nFilePath = join('root', '_i18n', 'i18n.csv');
            const doesExistSpy = jest.spyOn(utils, 'doesExist').mockResolvedValue(true);
            const readFileSpy = jest.spyOn(utils, 'readFile').mockResolvedValue('');
            const writeFileSpy = jest.spyOn(utils, 'writeFile').mockResolvedValue();
            // act
            const result = await tryAddCsvTexts(env, path, entries);
            // assert
            expect(result).toEqual(true);
            expect(doesExistSpy).toHaveBeenNthCalledWith(1, csvI18nFilePath);
            expect(readFileSpy).toHaveBeenNthCalledWith(1, csvI18nFilePath, undefined);
            expect(writeFileSpy).toHaveBeenNthCalledWith(1, csvI18nFilePath, 'key;en\nNewKey;New Value\n', undefined);
        });
        test('add to existing .csv file - mem-fs-editor', async () => {
            // arrange
            const memFs = create(createStorage());
            const csvI18nFilePath = join('root', '_i18n', 'i18n.csv');
            const doesExistSpy = jest.spyOn(utils, 'doesExist').mockResolvedValue(true);
            const readFileSpy = jest.spyOn(utils, 'readFile').mockResolvedValue('');
            const writeFileSpy = jest.spyOn(utils, 'writeFile').mockResolvedValue();
            // act
            const result = await tryAddCsvTexts(env, path, entries, memFs);
            // assert
            expect(result).toEqual(true);
            expect(doesExistSpy).toHaveBeenNthCalledWith(1, csvI18nFilePath);
            expect(readFileSpy).toHaveBeenNthCalledWith(1, csvI18nFilePath, memFs);
            expect(writeFileSpy).toHaveBeenNthCalledWith(1, csvI18nFilePath, 'key;en\nNewKey;New Value\n', memFs);
        });
    });
});
