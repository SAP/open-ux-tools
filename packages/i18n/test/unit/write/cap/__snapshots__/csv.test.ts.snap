// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csv add new i18n entries to csv file empty bundle 1`] = `
"key;fallback
key;value
"
`;

exports[`csv add new i18n entries to csv file empty fallback bundle 1`] = `
"key;fallback
key;value
"
`;

exports[`csv add new i18n entries to csv file empty file 1`] = `
"key;fallback
key;value
"
`;

exports[`csv add new i18n entries to csv file fallback bundle with multiple values 1`] = `
"key;fallback
key;value
key2;value2
"
`;

exports[`csv add new i18n entries to csv file multiple bundles with values 1`] = `
"key;en;de;fallback
key;e;d;
key;;;value
key2;;;value2
"
`;

exports[`csv add new i18n entries to csv file multiple bundles with values, but no fallback bundle 1`] = `
"key;en;de;fallback
key;e;d;
key;;;value
key2;;;value2

"
`;
